ProjFS.Mac/PrjFSLib/PrjFSLib.cpp

```diff
@@ -533,10 +533,31 @@ PrjFS_Result PrjFS_DeleteFile(
     }
 
     // TODO(Mac): Ensure that races with hydration are handled properly
-    // TODO(Mac): Ensure file is not full before proceeding
     
     char fullPath[PrjFSMaxPath];
     CombinePaths(s_virtualizationRootFullPath.c_str(), relativePath, fullPath);
+    
+    struct stat path_stat;
+    stat(fullPath, &path_stat);
+    if (!(S_ISREG(path_stat.st_mode) || S_ISDIR(path_stat.st_mode)))
+    {
+        // Only files and directories can be deleted with PrjFS_DeleteFile
+        // Anything else should be treated as a full
+        *failureCause = PrjFS_UpdateFailureCause_FullFile;
+        return PrjFS_Result_EVirtualizationInvalidOperation;
+    }
+    
+    if (S_ISREG(path_stat.st_mode))
+    {
+        // TODO(Mac): Determine if we need a similar check for directories as well
+        PrjFSFileXAttrData xattrData = {};
+        if (!GetXAttr(fullPath, PrjFSFileXAttrName, sizeof(PrjFSFileXAttrData), &xattrData))
```

Comment:
Speaking of races, I think we need to grab the file lock here so that we're not deleting at the same time as transitioning to full.

Reply:
As discussed offline, because the file modified notification is not a vnode callback, we can't synchronize converting a file to full and updating a placeholder.   However, the same race exists with git (without VFS4G), and so for now we can stick with the current changes and not add in locking.